File: GeometricObjects/PencilSource.py
from Vector import *
import random as r

r.seed(1)
class PencilSource:
    def __init__(self, start: Vector, end: Vector):
        """Sweeps along the line connecting start and end, 
        producing vectors starting along the line with uniformly random directions.
        Call getEmissionRays(n, numcores=2) to get n vectors produced, organized in numcores lists.

        Args:
            start (Vector): Start point of pencil
            end (Vector): End point of pencil
        """
        self.start = start
        self.end = end

    def randomSphericalVector(self):
        theta = 2*math.pi*r.random()
        phi = math.pi*r.random()
        sinphi = math.sin(phi)
        return Vector(sinphi*cos(theta), sinphi*sin(theta), cos(phi))
    
    def getEmissionRays(self, n, numcores = 1)-> list[list[tuple[Vector, Vector]]]:
        """Get n emission rays across numcores cores

        Args:
            n (_type_): _description_
            numcores (int, optional): _description_. Defaults to 1.

        Returns:
            list[list[tuple[Vector, Vector]]]: List sets of vectors divided for cores. Each set is a list of pairs of vectors
        """
        add_vec = (self.end - self.start ) / n
        out = []
        #print(f"{add_vec=}")
        st_i = 0    
        end_i = n//numcores
        for core in range(numcores):
            #print(f"{core=}")
            core_vecs = []
           # print(f"{st_i=}, {end_i=}")
            for i in range(st_i, end_i):
                #print(f"{n=}")
                core_vecs.append((self.start + i*add_vec, self.randomSphericalVector() ))
            st_i = end_i
            end_i += n // numcores
            out.append(core_vecs)
        out[-1].append((self.end, self.randomSphericalVector()))
        return out
    
class PencilSourceDebug:
    def __init__(self, start: Vector, end: Vector):
        """Sweeps along the line connecting start and end, 
        producing vectors starting along the line with uniformly random directions.
        Call getEmissionRays(n, numcores=2) to get n vectors produced, organized in numcores lists.

        Args:
            start (Vector): Start point of pencil
            end (Vector): End point of pencil
        """
        self.start = start
        self.end = end

    def randomSphericalVector(self):
        return Vector(1, 0, 0)
    
    def getEmissionRays(self, n, numcores = 1)-> list[list[tuple[Vector, Vector]]]:
        """Get n emission rays across numcores cores

        Args:
            n (_type_): _description_
            numcores (int, optional): _description_. Defaults to 1.

        Returns:
            list[list[tuple[Vector, Vector]]]: List sets of vectors divided for cores. Each set is a list of pairs of vectors
        """
        add_vec = (self.end - self.start ) / n
        out = []
        #print(f"{add_vec=}")
        st_i = 0    
        end_i = n//numcores
        for core in range(numcores):
            #print(f"{core=}")
            core_vecs = []
           # print(f"{st_i=}, {end_i=}")
            for i in range(st_i, end_i):
                #print(f"{n=}")
                core_vecs.append((self.start + i*add_vec, self.randomSphericalVector() ))
            st_i = end_i
            end_i += n // numcores
            out.append(core_vecs)
        out[-1].append((self.end, self.randomSphericalVector()))
        return out

File: rust/src/pencilsource.rs
use crate::vector::{Vector, Vector2};
use rand::Rng;
use std::f32::consts::PI;
use std::ops::{Add, Div, Sub};
pub struct PencilSource {
    pub start: Vector,
    pub end: Vector,
}

impl PencilSource {
    pub fn new(start: Vector, end: Vector) -> Self {
        PencilSource { start, end }
    }

    pub fn random_spherical_vector(&self) -> Vector {
        let theta = 2.0 * PI * rand::thread_rng().gen::<f32>();
        let phi = PI * rand::thread_rng().gen::<f32>();
        let sin_phi = phi.sin();
        Vector {
            x: sin_phi * theta.cos(),
            y: sin_phi * theta.sin(),
            z: phi.cos(),
        }
    }

    pub fn get_emission_rays(&self, n: usize, num_cores: usize) -> Vec<Vec<(Vector, Vector)>> {
        let add_vec = (self.end - self.start) / n as f32;
        let mut out = Vec::new();
        let mut st_i = 0;
        let mut end_i = n / num_cores;
        for _ in 0..num_cores {
            let mut core_vecs = Vec::new();
            for i in st_i..end_i {
                core_vecs.push((
                    self.start + add_vec * (i as f32),
                    self.random_spherical_vector(),
                ));
            }
            st_i = end_i;
            end_i += n / num_cores;
            out.push(core_vecs);
        }
        out.last_mut()
            .unwrap()
            .push((self.end, self.random_spherical_vector()));
        out
    }
}

pub struct PencilSourceDebug {
    start: Vector,
    end: Vector,
}

impl PencilSourceDebug {
    pub fn new(start: Vector, end: Vector) -> Self {
        PencilSourceDebug { start, end }
    }

    pub fn random_spherical_vector(&self) -> Vector {
        Vector {
            x: 1.0,
            y: 0.0,
            z: 0.0,
        }
    }

    pub fn get_emission_rays(&self, n: usize, num_cores: usize) -> Vec<Vec<(Vector, Vector)>> {
        let add_vec = (self.end - self.start) / n as f32;
        let mut out = Vec::new();
        let mut st_i = 0;
        let mut end_i = n / num_cores;
        for _ in 0..num_cores {
            let mut core_vecs = Vec::new();
            for i in st_i..end_i {
                core_vecs.push((
                    self.start + add_vec * (i as f32),
                    self.random_spherical_vector(),
                ));
            }
            st_i = end_i;
            end_i += n / num_cores;
            out.push(core_vecs);
        }
        out.last_mut()
            .unwrap()
            .push((self.end, self.random_spherical_vector()));
        out
    }
}

fn main() {
    // Example usage
    let start = Vector {
        x: 0.0,
        y: 0.0,
        z: 0.0,
    };
    let end = Vector {
        x: 1.0,
        y: 1.0,
        z: 1.0,
    };

    let pencil_source = PencilSource::new(start, end);
    let emission_rays = pencil_source.get_emission_rays(10, 2);
    println!("{:?}", emission_rays);

    let pencil_source_debug = PencilSourceDebug::new(start, end);
    let emission_rays_debug = pencil_source_debug.get_emission_rays(10, 2);
    println!("{:?}", emission_rays_debug);
}
